components/content/content-editor/controls/mention/autocomplete/controls.ts

```diff
@@ -0,0 +1,254 @@
+import AppScrollScroller from '../../../../../scroll/scroller/scroller.vue';
+import AppUserAvatarImg from '../../../../../user/user-avatar/img/img.vue';
+import { User } from '../../../../../user/user.model';
+import AppUserVerifiedTick from '../../../../../user/verified-tick/verified-tick.vue';
+import { MentionCache, MentionCacheUser } from '../../../../mention-cache';
+import { ContentEditorService } from '../../../content-editor.service';
+import { ContentEditorSchema } from '../../../schemas/content-editor-schema';
+
+@Component({
+	components: {
+		AppUserAvatarImg,
+		AppUserVerifiedTick,
+		AppScrollScroller,
+		AppLoading,
+	},
+})
+export default class AppContentEditorControlsMentionAutocompleteControls extends Vue {
+	@Prop(Object)
+	view!: EditorView<ContentEditorSchema>;
+	@Prop(Number)
+	stateCounter!: number;
+	@Prop(Number)
+	canShow!: number;
+
+	// We cache prior search results so we don't need to query backend again.
+	private static searchResults: { [k: string]: User[] } = {};
+
+	suggestion = ''; // Currently active suggestion.
+	users: MentionCacheUser[] = [];
+	visible = false;
+	top = 'auto';
+	bottom = 'auto';
+	left = 'auto';
+
+	selectedIndex = 0;
+	isListening = false; // If we are listening to the document keydown event, to be able to unbind it later.
+	inverted = false; // If the list is inverted due to screen size constraints. It shows above the control instead of below.
+	remoteSuggestionTimeout?: NodeJS.Timer; // Timeout between requests to search backend
+	isLoading = false; // Loading more users from backend
+
+	readonly Screen = Screen;
+
+	$refs!: {
+		container: HTMLElement;
+		list: HTMLDivElement;
+	};
+
+	get displayUsers() {
+		if (this.inverted) {
+			return this.users.reverse();
+		}
+		return this.users;
+	}
+
+	get searchResultsCache() {
+		// just a shorthand
+		return AppContentEditorControlsMentionAutocompleteControls.searchResults;
+	}
+
+	mounted() {
+		this.update();
+		document.addEventListener('keydown', this.onKeyDown);
+		this.isListening = true;
+	}
+
+	destroyed() {
+		if (this.isListening) {
+			document.removeEventListener('keydown', this.onKeyDown);
+			this.isListening = false;
+		}
+	}
+
+	@Watch('canShow')
+	@Watch('stateCounter')
+	private update() {
+		if (this.canShow && this.view instanceof EditorView) {
+			const state = this.view.state;
+			const node = ContentEditorService.getSelectedNode(this.view.state);
+
+			if (node !== null && node.isText) {
+				const slice = state.doc.slice(0, state.selection.from);
+				const text = ContentEditorService.getFragmentText(slice.content);
+				const matches = /@([\w_-]+)$/i.exec(text);
+
+				if (matches && matches.length >= 2) {
+					const start = this.view.coordsAtPos(state.selection.from);
+
+					const box = this.$refs.container.offsetParent.getBoundingClientRect();
+
+					// If the text control is more than 50% down the page, open the control above ("inverted")
+					const relativeYPos = box.top / Screen.height;
```

Comment:
There seems to be all sorts of wrong with this. The box is positioned very weirdly in some cases, we should talk about this.